### PR TITLE
Add customizations for more flexibility, a bit of refactoring

### PR DESCRIPTION
**Pull request by a13**
Hi, see the description of changes in comments :)

### CODE REVIEW

File: babashka.el [hunk 1]
```diff
@@ -50,51 +76,66 @@
...
   "Run a shell COMMAND in a DIRECTORY and display output in OUTPUT-BUFFER."
   (let ((default-directory directory))
-    (async-shell-command command output-buffer)))
+    (funcall babashka-async-shell-command command)))
 
 (defun babashka--locate-bb-edn (&optional dir)
   "Recursively search upwards from DIR for bb.edn file."
-  (if-let ((found (locate-dominating-file (or dir default-directory) "bb.edn")))
-      (concat found "bb.edn")))
+  (when-let ((found (locate-dominating-file (or dir default-directory) "bb.edn")))
```
Comment (a13) [inline babashka.el:86]: it's more idiomatic to use `when-let` when there's a single branch

File: babashka.el [hunk 2]
```diff
@@ -37,8 +37,34 @@
...
 ;;; Code:
 (require 'parseedn)
+(require 'seq)
 (require 'project)
 
+(defgroup babashka nil
+  "Babashka Tasks Interface"
+  :group 'external)
+
+(defcustom babashka-async-shell-command #'async-shell-command
```
Comment (a13) [inline babashka.el:47]: in theory, we can use this to run tasks outside of Emacs as well, like terminal emulators, tmux etc

File: babashka.el [hunk 3]
```diff
@@ -50,51 +76,66 @@
...
+  "Annotate S using current completiong table."
+  (when-let ((item (assoc s minibuffer-completion-table)))
+    (concat " " (cdr item))))
+
+(defun babashka--tasks-to-annotated-names (tasks)
+  "Convert TASKS to annotated alist."
+  (let (results)
+    (maphash (lambda (key value)
+               (let ((task-name (symbol-name key)))
+                 (unless (string-prefix-p ":" task-name)
```
Comment (a13) [inline babashka.el:110]: we don't want to see `:init` `:require` etc here

File: babashka.el [hunk 4]
```diff
@@ -50,51 +76,66 @@
      (insert-file-contents file-path)
      (buffer-string))))
 
-(defun babashka--run-shell-command-in-directory (directory command &optional output-buffer)
+(defun babashka--run-shell-command-in-directory (directory command)
```
Comment (a13) [inline babashka.el:79]: Also, I don't see any usage of `output-buffer` arg here, so I decided to remove it for unification

File: babashka.el [hunk 5]
```diff
@@ -50,51 +76,66 @@
...
 
+
 (defun babashka--escape-args (s)
   "Shell quote parts of the string S that require it."
   (mapconcat #'shell-quote-argument (split-string s) " "))
 
+(defun babashka--annotation-function (s)
+  "Annotate S using current completiong table."
+  (when-let ((item (assoc s minibuffer-completion-table)))
+    (concat " " (cdr item))))
```
Comment (a13) [inline babashka.el:103]: I'd prefer something more fancy, like `ivy-rich` etc, but this one is more generic solution.

**Comment by licht1stein** [comment]
You're seriously good :) thanks!

**Pull request merged**